### DISCUSSION TITLE
Cast INumber to double and back

### DISCUSSION THREAD

**Discussion by arnirichard** [dotnet/csharplang#7069] [General] [answered]
I am creating a filter that I want to be able to filter array of `INumber<T>`.

However the internal state of the filter must have high precision using doubles.

How can I design the filter that has ability to convert `INumber<T>` to double and back?

For example casting to double could be done as folllows.

```
private double ToDouble<K>(K? t)
{
    if (t is int i)
        return i;
    if (t is byte b)
        return b;
    if (t is double d)
        return d;
    if (t is float f)
        return f;
    if (t is short s)
        return s;
    if (t is uint u)
        return u;
    if (t is decimal de)
        return (double)de;
    return 0;
}
```

But how would I be able to cast the filtering result from double to INumber<T>?

**Comment by Joe4evr** [comment, marked as answer]
Use `double.CreateChecked(num)` or `CreateSaturating` or `CreateTruncating` depending on your needs.

**Reply by arnirichard** [reply]
Thank you so much!

**Comment by TonyValenti** [comment]
If you need high precision, you might also consider looking at the decimal data tight as it is better for high precision small numbers.

**Comment by arnirichard** [comment]
@TonyValenti Thanks, that is a good tip. Actually I made filter that can also filter array of bytes, so I did not want to keep the internal state of the filter in bytes. So now I can use double/decimal for the filter which has bytes input and output.
